scipy/sparse/linalg/isolve/iterative.py

```diff
@@ -148,70 +151,73 @@ def combine(fn):
                )
 @non_reentrant()
 def bicg(A, b, x0=None, tol=1e-5, maxiter=None, M=None, callback=None, atol=None):
-    A,M,x,b,postprocess = make_system(A, M, x0, b)
-
-    n = len(b)
+    assert A.shape[0] == A.shape[1] and A.shape[0] == len(b), "The size of the matrix and the right-hand side does not match."
     if maxiter is None:
-        maxiter = n*10
-
-    matvec, rmatvec = A.matvec, A.rmatvec
-    psolve, rpsolve = M.matvec, M.rmatvec
-    ltr = _type_conv[x.dtype.char]
-    revcom = getattr(_iterative, ltr + 'bicgrevcom')
-
-    get_residual = lambda: np.linalg.norm(matvec(x) - b)
-    atol = _get_atol(tol, atol, np.linalg.norm(b), get_residual, 'bicg')
-    if atol == 'exit':
-        return postprocess(x), 0
-
-    resid = atol
-    ndx1 = 1
-    ndx2 = -1
-    # Use _aligned_zeros to work around a f2py bug in Numpy 1.9.1
-    work = _aligned_zeros(6*n,dtype=x.dtype)
-    ijob = 1
-    info = 0
-    ftflag = True
-    iter_ = maxiter
-    while True:
-        olditer = iter_
-        x, iter_, resid, info, ndx1, ndx2, sclr1, sclr2, ijob = \
-           revcom(b, x, work, iter_, resid, info, ndx1, ndx2, ijob)
-        if callback is not None and iter_ > olditer:
+        maxiter = 10000
+    # type judgment and conversion
+    dtype = A.dtype
+    if dtype == int:
+        dtype = float
+        A = A.astype(dtype)
+        if b.dtype == int:
+            b = b.astype(dtype)
+    A, M, x, b, postprocess = make_system(A, M, x0, b)
+    # judge if b is a zero vector
+    if np.linalg.norm(b) == 0.:
+        x = b.copy()
+        return (postprocess(x), 0)
+    At = A.H
+    Mt = M.H
+    r = b - A.matvec(x)
+    z = M.matvec(r)
+    rhat = z.copy()
+    z_hat = Mt.matvec(rhat)
+    d_old = np.inner(rhat.conjugate(), z)
+    p = r.copy()
+    p_hat = z.copy()
+    beta = 0.
+    r0norm = np.linalg.norm(r)
+    if r0norm == 0.:
+        return (postprocess(x), 0)
+    if atol is not None:
+        if atol == 'legacy':
+            if r0norm <= tol:
+                return postprocess(x), 0
+        else:
+            tol = max(float(atol)/r0norm, tol)
+    for iter in range(maxiter):
+        p = z + beta * p
+        p_hat = z_hat + beta.conjugate() * p_hat
+        Ap = A.matvec(p)
+        phattAp = np.inner(p_hat.conjugate(), Ap)
+        if phattAp == 0.:
+            print("Warning: divide by zero (in step {})".format(iter+1))
```

Comment:
I would use the `warning` module here to explicitly raise a warning instead of the print function.

Reply:
I'll use `raise` to change it. Thanks